urlbc3e98d97e_jmhorjus_EFCP_Implementation_tgz-pJ8-UDPConnector_UdpConnectorTestJ8/src/UDPConnector/UdpConnector.java

```diff
@@ -0,0 +1,316 @@
+    {
+        //System.out.print("UdpConnector.AddReceiveNotify: Enter.\n");
+        boolean retVal = false;
+        synchronized(m_notifyOnReceive){         
+            retVal = m_notifyOnReceive.add(notifyMe);
+        }
+        return retVal;
+    }
+    
+    /// Non-blocking receive.
+    /// Returns either a list of available packets or null.
+    /// maxBlockingTimeInMs: may be either zero (for non-blocking receive) 
+    /// or a positive number of milliseconds - the longest you are willing to 
+    /// wait for at least one packet to be available.
+    @Override
+    public List<byte[]> Receive(int maxBlockingTimeInMs)
+    {
+        // Make sure receive thread is started.
+        StartReceiveThread();
+
+        
+        List<byte[]> retVal = new LinkedList<>(); 
+        
+        synchronized(m_receiveBuffer)
+        {
+            // If no packets are ready, wait up to maxBlockingTime for a signal.
+            if (m_packetsReady == 0 && maxBlockingTimeInMs > 0)
+            {
+                //System.out.print("UdpConnector.Receive: waiting on receive...\n");
+                try {
+                    m_receiveBuffer.wait(maxBlockingTimeInMs);
+                } catch (Exception ex) { 
+                    System.out.print("**ERROR** Exception waiting for next packet in Receive: \""+ex.getMessage()+"\"\n");
+                }
+                System.out.print("UdpConnector.Receive: Done waiting on receive packetsReady=" + m_packetsReady + "\n");
+            }
+            else
+            {
+                //System.out.print("UdpConnector.Receive: NOT waiting. Packetsready=" + m_packetsReady + "\n");
+            }
+        }
+        
+        while (m_packetsReady > 0)
+        {
+            synchronized(m_receiveBuffer)
+            {
+                //System.out.print("UdpConnector.Receive: adding:\"" + new String(m_receiveBuffer[m_receiveBufferReadIndex]) + "\"\n");
+                
+                retVal.add(m_receiveBuffer[m_receiveBufferReadIndex]);
+                
+                m_receiveBufferReadIndex = 
+                        (m_receiveBufferReadIndex+1)%m_receiveBuffer.length;   
+                --m_packetsReady;
+            }
+        }
+        return retVal;
+    }
+    
+    
+    /// Sending is much easier than Receiving!
+    /// Just send it and don't care!
+    @Override
+    public boolean Send(String sendString) throws Exception
+    {   
+        return Send(sendString.getBytes());
+    }
+    
+    @Override
+    public boolean Send(byte[] sendBuffer) throws Exception
+    {     
+        //System.out.print("UdpConnection.Send: Enter.  buf="+ new String(sendBuffer) +" destPort="+ m_peerPort +".\n");
+        if (!this.m_isPeerSet)
+        {
+            System.out.print("ERROR: Peer not set; cannot send.");
+            return false;
+        }
+        
+        DatagramSocket sendSocket = new DatagramSocket();
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
This line of code might contain a resource leak. Resource leaks can cause your system to slow down or crash.

**Fix**
Consider closing the following resource: *sendSocket*. Execution paths that do not contain closure statements were detected. Ensure *sendSocket* is closed in a try-finally block or declared in a try-with-resources block.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).